### PR TITLE
[POC] ensure migration ability with resource quota

### PR DESCRIPTION
**Pull request by Barakmor1**
**Reviewer Checklist**

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

Please note that the PR is not complete as we still need to add tests.
Additionally, we have identified an issue with handling web-hook requests simultaneously. This can result in conflicts, such as the scenario where there is limited spare capacity for a VMI migration and multiple Pod creation requests are submitted at the same time. The `MigrationCapacityValidatorWebhook` may approve one of the Pod requests without considering resources needed for the other, and the Quota admission controller may approve both requests without knowledge of the migration resources. As a result, there may not be enough resources left for the migration.

Possible solutions To address this issue: 
- The first solution is to handle the creation of ReplicaSets, Deployments, and DaemonSets in addition to Pods, which may decrease the likelihood of conflicts. 
- The second solution is to allow the web-hook to mark the Pod and then add another controller to validate the Pod after creation. If a conflict is detected, the controller will recreate the Pod, and if not, it will delete the mark on the Pod but than the user will wonder why did the pod wasn't created.

- [ ] PR Message
- [ ] Commit Messages
- [ ] How to test
- [ ] Unit Tests
- [ ] Functional Tests
- [ ] User Documentation
- [ ] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly
  
**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by nunnatsa** [comment]
This is a huge amount of code that is not relevant to HCO, including many irrelevant dependencies.

It will make HCO harder to built and be maintained.

**Comment by nunnatsa** [comment]
Also, please rebase

File: pkg/components/components.go [hunk 1]
```diff
@@ -463,6 +463,16 @@ func GetClusterPermissions() []rbacv1.PolicyRule {
...
+		{
+			APIGroups: stringListToSlice("kubevirt.io"),
+			Resources: stringListToSlice("virtualmachineinstancemigrations"),
+			Verbs:     stringListToSlice("list", "watch"),
+		},
+		{
+			APIGroups: stringListToSlice("kubevirt.io"),
+			Resources: stringListToSlice("virtualmachineinstances"),
+			Verbs:     stringListToSlice("list", "watch"),
+		},
```
[the remark this answers is not in the document]
Comment (Barakmor1) [reply pkg/components/components.go]: Done

**Comment by Barakmor1** [comment]
> This is a huge amount of code that is not relevant to HCO, including many irrelevant dependencies.
> 
> It will make HCO harder to built and be maintained.

To validate whether a Pod can be created based on the resource quota rule, we are importing several modules to use the CheckRequest function provided by Kubernetes. While it is possible to rewrite this function from scratch without relying on Kubernetes, it would require significant effort and ongoing support, which could lead to the introduction of bugs in the future and require additional resources.

Despite the potential downsides of importing all this code into HCO, I believe that the benefits of using the CheckRequest function outweigh the risks.

**Comment by Barakmor1** [comment]
@iholder101 @vladikr @dankenigsberg @fabiand

**Comment by nunnatsa** [comment]
> > This is a huge amount of code that is not relevant to HCO, including many irrelevant dependencies.
> 
> > 
> 
> > It will make HCO harder to built and be maintained.
> 
> 
> 
> To validate whether a Pod can be created based on the resource quota rule, we are importing several modules to use the CheckRequest function provided by Kubernetes. While it is possible to rewrite this function from scratch without relying on Kubernetes, it would require significant effort and ongoing support, which could lead to the introduction of bugs in the future and require additional resources.
> 
> 
> 
> Despite the potential downsides of importing all this code into HCO, I believe that the benefits of using the CheckRequest function outweigh the risks.

The dependencies issue id just part of the bigger problem. The whole thing does not belong to hco. HCO should not know about vmi, for example. This implementation breaks several  design principles. e.g. it creates coupling when it shouldn't.

**Pull request closed without merging**